src/daemonizer/posix_fork.cpp

```diff
@@ -21,15 +21,42 @@
 namespace posix {
 
 namespace {
-  void quit(std::string const & message)
+  void quit(const std::string & message)
   {
     LOG_ERROR(message);
     throw std::runtime_error(message);
   }
 }
 
-void fork()
+void fork(const std::string & pidfile)
 {
+  // If a PID file is specified, we open the file here, because
+  // we can't report errors after the fork operation.
+  // When we fork, we close thise file in each of the parent
+  // processes.
+  // Only in the final child process do we write the PID to the
+  // file (and close it).
+  std::ofstream pidofs;
+  if (! pidfile.empty ())
+  {
+	int oldpid;
+    std::ifstream pidrifs;
+    pidrifs.open(pidfile, std::fstream::in);
+    if (! pidrifs.fail())
+    {
+	  if (pidrifs >> oldpid && oldpid > 1 && kill(oldpid, 0) == 0)
+      {
+        quit("PID file " + pidfile + " already exists and the PID therein is valid");
```

Comment:
As mentioned before. This logic to quit just because there is another running daemon makes no sense. Just log an error to warn the user and then leave it to the user to decide what to do.